### ISSUE TITLE
GSoC Summer Project: Implement full Hadoop Compatibility Layer for Stratosphere

### ISSUE THREAD

**Issue by rmetzger** [stratosphere/stratosphere#838]
This is a meta issue for tracking @atsikiridis progress with implementing a full Hadoop Compatibliltiy Layer for Stratosphere.

Some documentation can be found in the Wiki: https://github.com/stratosphere/stratosphere/wiki/%5BGSoC-14%5D-A-Hadoop-abstraction-layer-for-Stratosphere-(Project-Map-and-Notes)

As well as the project proposal: https://github.com/stratosphere/stratosphere/wiki/GSoC-2014-Project-Proposal-Draft-by-Artem-Tsikiridis

Most importantly, there is the following **schedule**:

_19 May - 27 June (Midterm)_

1) Work on the Hadoop tasks, their Context and the mapping of Hadoop's Configuration to the one of Stratosphere. By successfully bridging the Hadoop tasks with Stratosphere, we already cover the most basic Hadoop Jobs. This can be determined by running some popular Hadoop examples on Stratosphere (e.g. WordCount, k-means, join) (4 - 5 weeks)

2) Understand how the running of these jobs works (e.g. command line interface) for the wrapper. Implement how will the user run them. (1 - 2 weeks).

_27 June - 11 August_

1) Continue wrapping more "advanced" Hadoop Interfaces (Comparators, Partitioners, Distributed Cache etc.) There are quite a few interfaces and it will be a challenge to support all of them. (5 full weeks)

2) Profiling of the application and optimizations (if applicable)

_11 August - 18 August_

Write documentation on code, write a README with care and add more unit-tests. (1 week)

**Comment by rmetzger** [comment]
@atsikiridis created a first pull request for this https://github.com/stratosphere/stratosphere/pull/833

**Comment by rmetzger** [comment]
I suggest that the merge the work of @atsikiridis into the `master` once we have a working wrapper for Mappers and Reducers.

The next point where i suggest a merge is when we have scripts to run MR jobs on Stratosphere, without changing a single line of code.

**Comment by uce** [comment]
Thanks for the update. I think it's a really cool summer project. :-)

**Comment by atsikiridis** [comment]
Hello!

@uce I agree, thanks for the interest :-)

A first very interesting week has passed and it's time for my first report. I'll be doing a weekly report of what was been done in the previous week and the plans for the next week and of course mentioning anything else necessary during the week. Please, feel free to submit any comments you would like, as they will be very helpful!

**What was done this week?**

**tl;dr version**

Wrappers for Hadoop Mappers and Reducers (Combiners too). A WordCount works! Skeleton for abstraction layer client (a `HadoopOnStratosphereJobClient` that is - more soon)

**Slightly longer version**

In the community bonding period I tried to understand the project as better as possible so that there would be no surprises. I was aware that there is a new Java API released, how Hadoop components would be mapped etc. Moreover, I started experimenting with the code in hadoopcompatibility and was really happy to make a sloppy but working WordCount. So why not clean it up and make a glorious entrance to GSoC with a PR ( #833 ) on the first day :P ? 

I did, but well since it's on the soon to be deprecated Records API it wasn't really useful. So, I ported on @twalthr 's #777 where you can see a hadoopcompatibility with the new Java API. The result of this is in this [branch](https://github.com/atsikiridis/stratosphere/tree/HadoopCompatibilityJAPIReady). So the WordCount test passes on this branch which is cool! It means the wrappers are working for a pretty general case (different input output types, combiner etc.)  There were some challenges though. See #842 (I still do a couple of hacks for that) and #860 ( the latter wasn't an issue after all, I just got confused but @fhueske got to the rescue).  And the `ReducerUnwrappingIterator` needs unit-tests urgently as it has been a source of confusion and bugs.

In the meantime, I've started working a `HadoopOnStratosphereJobClient` (I'll publish the branch latter today). This is my idea of the user's API.

```
jobConf ... // a jobConf with all Hadoop params for the job
HadoopOnStratosphereJobClient.runJob(jobConf);
```

This aspires to be the only line of code the user will write. In the following days I'll have a first version of that with a test. However, I believe it will be important to allow more that one jobConfs to be passed to have one PACT.

So this:

```
HadoopOnStratosphereJobClient jobClient = new HadoopOnStratosphereJobClient();
jobClient.submitJob(jobConf1);
jobClient.submitJob(jobConf2);
....
jobClient.runJobs();
```

This will be advantageous to chained jobs and their translations to PACT. More on that this week. Any suggestions?

**What about next week?**
- Unit tests for `ReduceUnwrappingIterator` and some cleaning up. @rmetzger Should I do a new PR?
- Make the current work on the JobClient public (that will be up today)
- More work on the jobClient. Make it standarized. Try many example hadoop Job drivers on Stratosphere (not just wordCount). Do they work with just this line of code?

**Comment by fhueske** [comment]
Hi Artem,

thanks for the update and good progress! :-)

Regarding the subclassing of the HadoopFunctionWrappers to get the generics
and the combinable annotation fixed: We should overcome this limitation.
@twalthr will implement the ResultTypeQueryable interface that lets you
explicitly return the result type of a function (so the typeextractor does
not need to look this up). The combiner can also be enabled by calling
setCombinable(true) on the ReduceGroupOperator which is returned by the
.reduceGroup() call. Maybe we need another hook here too.

Fabian

2014-05-26 5:02 GMT+02:00 Artem Tsikiridis notifications@github.com:

**Comment by atsikiridis** [comment]
So, map and reduce tasks are done with no hacks on the new Java API ( thanks to #868 ) and some other minor stuff are done in a more sane way.

here:
https://github.com/atsikiridis/stratosphere/tree/HadoopCompatibilityJAPIReady

I have prepared more tests (2 example jobs and some Iterator cases) and they will be added later today. Moreover, I run a WordCount via the first version of the driver I was talking about on Sunday. Yeah the one with too long name ( :( )  `HadoopOnStratosphereJobClient` . it'll be presentable definitely in the coming days( hopefully tomorrow or the day after).

Cheers!

**Comment by fhueske** [comment]
Nice! :smile:
Please watch #868 and open a PR against Master once it has been merged.

**Comment by rmetzger** [comment]
Very cool. I've checked out the branch and the tests successfully ran. Also, the code looks nice.
I was wondering if it is necessary to set all these configuration values:

``` java
hadoopJobConf.setMapperClass(TestTokenizeMap.class);

//In this job there is no need to specify out types for map output as they are the same
hadoopJobConf.setOutputKeyClass(Text.class);
hadoopJobConf.setOutputValueClass(LongWritable.class);

final DataSet<Tuple2<Text, LongWritable>> words = text.flatMap( new HadoopMapFunction<LongWritable,Text,Text, LongWritable>(hadoopJobConf));
```

because you're saying almost the same with the `text.flatMap()`call: The types, the operation (map), just the mapper class itself is missing.

**Comment by atsikiridis** [comment]
Hello @rmetzger .Thanks!

I have to give the output values to the wrappers somehow (above is one way) due to type erasure. The statements setOutput etc. are necessary in that case, as otherwise Hadoop considers Input types same as output types ( by using `getOutputKeyClass()` to retrieve the values  for example).

Of course I can avoid it and just pass the key-value classes to the constructor (as is already being done in InputFormat)... Do you think it's better? In this case, yeah, we pass just the mapper class to the wrapper and not the whole JobConfig.

**Comment by rmetzger** [comment]
Mh, okay, I see. Then lets keep it this way for now, since changing it is really not difficult.
Users working with the Map-Reduce classes are used to the HadoopJobConf stuff anyways.

**Comment by atsikiridis** [comment]
Hello,

https://github.com/atsikiridis/stratosphere/blob/hadoop-driver/stratosphere-addons/hadoop-compatibility/src/main/java/eu/stratosphere/hadoopcompatibility/mapred/StratosphereHadoopJobClient.java

In this branch I run a job with no stratoshere code other than the `StratosphereHadoopJobClient.runJob(conf);` conmand in the end. There's a wordcount test passing. Well, I don't think it's general enough yet and I' m still working on the mapping of parallelism and there are some hacks. Moreover, I must have a test case with two map  tasks, what if there is no reducer, accesing stats of the running job etc. I have tried some stuff and they will be up by the end of this week.

**Comment by rmetzger** [comment]
Cool.
So you're setting the DOP just to 1. I'm not sure if it will be a problem, but Stratosphere and MapReduce behave differently regarding the number of mappers and reducers.
In MR, its usually the number of files, in Stratosphere, its the DOP (or the number of files for unsplittable files such as avro or deflate).

I'm not sure if the handling of the Combiner-Class is correct right now. You just say "combinable" if there is such a class, but you don't call it.

**Comment by atsikiridis** [comment]
@rmetzger Ahh yes you're right for the combinable. Thanks.

**Comment by atsikiridis** [comment]
Hello,
let me discuss a bit about the second week.

**What was done this week?**

**tl;dr version**

Mainly, one can run WordCount (and other jobs using only Mappers, Reducers and Combiners) without modifying the code of his driver. `StratosphereHadoopJobClient.runJob(conf).waitForCompletion();` is enough (check: https://github.com/atsikiridis/stratosphere/tree/hadoop-driver/stratosphere-addons/hadoop-compatibility/src/main/java/eu/stratosphere/hadoopcompatibility/mapred), you can try the tests . Some limitations. Moreover, worked on the `RunningJob` and did some minor refactoring to basic wrappers branch.

**Slightly longer version**

As stated in the project proposal, in this first phase (and allow me to assume a bit less, to focus more on adnvanced programming interfaces) we must have a full implementation of the basic programming interfaces and how a job is being run. While I would like to think that the first part is mainly done, (check https://github.com/atsikiridis/stratosphere/tree/HadoopCompatibilityJAPIReady) sometimes it can't be helped and stuff need to be changed to play well with how the driver looks like.  As this may occur again during the implementation I consider these closely tied for now.

First of all,  a whole hadoop configuratio is no longer passed to basic wrappers but just the class and the types of the key and value to be produced (e.g. mapper, Text, LongWritable).  This turned out to be better to support all combiner use cases which can be identified as the following:

1) Same Combiner as Reducer --> `setCombinable` is enough
2) Different Combiner  ---> an additional `GroupReduceOperator`
3) No Combiner at all --->Nothing additional :)

For hadoop we can assume there's always one mapper and one reducer as even if the user doesn't specify any of them identity functions will be called :) And that's how the driver works as well.

There are some test cases included, though they should probably be restructured.

And here's the tricky part of the week:we need to have the input key and value types (classes) of the `InputFormat`! So one would say, ok use the `TypeExtractor` like this:

`TypeExtractor.getParameterType(baseClass, clazz, pos)`

Unfortunately, for the `mapred` api there is no de facto InputFormat base class with generics. One could say it's `FileInputFormat<K,V>` for most but then we limit the support for InputFormats accepting most file-based only.

So I've come up with the following not really nice "solutions" (both have their weaknesses but work mostly):

1)  the tree of classes is traversed up until I find the first inputformat class hoping it is generic with <K,V> args. Like this:

```
    private Class<InputFormat<?,?>> getInputFormatBaseClass(Class<InputFormat> inputFormatClass) {

        Class sc = inputFormatClass.getSuperclass();
        Type scType = inputFormatClass.getGenericSuperclass();
        while(scType != null) {
            if (scType instanceof ParameterizedType && !(InputFormat.class.isAssignableFrom(sc.getSuperclass()))) {
                final int noOfTypes =  ((ParameterizedType) scType).getActualTypeArguments().length;
                if (noOfTypes >=2) {  //needs improvement TODO
                    break;
                }
            }
            sc = sc.getSuperclass();
            scType = ((Class) scType).getGenericSuperclass();
        }
        if (sc == null) {

        }
        return sc;
    }
```

Whick works if there is one such class ( in most cases there is one!). Another approach would be:

2)

```
    private Class[] getInputFormatClasses(InputFormat inputFormat, JobConf jobConf) throws IOException{
        final Class[] inputFormatClasses = new Class[2];
        final InputSplit firstSplit = inputFormat.getSplits(jobConf, 0)[0];
        final Reporter reporter = new HadoopDummyReporter();
        inputFormatClasses[0] = inputFormat.getRecordReader(firstSplit, jobConf, reporter).createKey().getClass();
        inputFormatClasses[1] = inputFormat.getRecordReader(firstSplit, jobConf, reporter).createValue().getClass();
        return inputFormatClasses;
    }
```

If there is an inputsplit yeah it will work. Maybe this does work for everything after all (really, can we have inputformats with no inputsplits?)...

In the very worse case, ask the key-value classes from the user violating the single line of code principle :( .

This is a tricky part and I'll try to come up with something better next week...

Then,  some work has been done  on `StratosphereRunningJob` which is a `RunningJob` You can get jobstatus and counters (accumulators) data from the running job via this class. It is important that we just extend theses classes and preserve the same API so that the user really won't have to change a thing in a more complicated job driver. So right now most i've replaced the points of entries to the hadoop cluster to be executed on stratosphere but the work is in half way as we don't have a mapping of counters to accumulators... And we need to check how to access the nephele cluster from the execution environment e.g. to kill a job. (if possible at all).

**What about next week?**
- I believe a realistic goal is to not have anything dummy in the current code. So counters --> accumulators, Reporter etc and this will result to a fully fledged `StratosphereRunningJob`. Test with more complex job drivers.
- Solve the inputFormat problem or check if solution 2 is general enough.
- With these we can have a full first version of a driver available for pull request. Then it would be nice to know that will the functionality for a basic driver is fully there we can incrementally add support for the more advanced programming interfaces (Partitioning, DistributedCache) and have a nice working thing everytime. Until we wrap everything :)

If anything wasn't clear or you would like to mention something else, please feel free :)
Cheers!

**Comment by atsikiridis** [comment]
About the input format issue: one other thing that can be done is to extend `HadoopInputFormat`  ( `new HadoopInputFormat(jobConf){}` ) and not require a class of key and value. Of course in this case we will not implement `ResultTypeQueryable` for inputformats. It's still a hack but covers everything really  So for now it's much better than the hideous stuff above so I'll keep it on my branch.

Cheers.

**Comment by atsikiridis** [comment]
Thanks @uce !

Yes then probably 2) is what I'll keep. However, it would be much much cleaner if we had a:

`TypeExtractor.getParameter(InputFormat.class, inputFormatClass, 0);` but make it work with an interface instead of requiring a baseclass. But I'm not sure how can we have this. I noticed in the comments of #895  that it may be possible (@twalthr mentions it).

**Comment by twalthr** [comment]
Yes, it's not a big thing to create a `TypeExtractor.getParameter()` which takes interfaces. I will implement it soon, we will need it for lambdas anyway.
